// @ethereumjs/block/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.valuesArrayToHeaderData = exports.getDifficulty = exports.genWithdrawalsTrieRoot = exports.genTransactionsTrieRoot = exports.genRequestsRoot = exports.executionPayloadFromBeaconPayload = void 0;
__exportStar(require("./block/index.js"), exports);
__exportStar(require("./consensus/index.js"), exports);
var from_beacon_payload_ts_1 = require("./from-beacon-payload.js");
Object.defineProperty(exports, "executionPayloadFromBeaconPayload", { enumerable: true, get: function () { return from_beacon_payload_ts_1.executionPayloadFromBeaconPayload; } });
__exportStar(require("./header/index.js"), exports);
var helpers_ts_1 = require("./helpers.js");
Object.defineProperty(exports, "genRequestsRoot", { enumerable: true, get: function () { return helpers_ts_1.genRequestsRoot; } });
Object.defineProperty(exports, "genTransactionsTrieRoot", { enumerable: true, get: function () { return helpers_ts_1.genTransactionsTrieRoot; } });
Object.defineProperty(exports, "genWithdrawalsTrieRoot", { enumerable: true, get: function () { return helpers_ts_1.genWithdrawalsTrieRoot; } });
Object.defineProperty(exports, "getDifficulty", { enumerable: true, get: function () { return helpers_ts_1.getDifficulty; } });
Object.defineProperty(exports, "valuesArrayToHeaderData", { enumerable: true, get: function () { return helpers_ts_1.valuesArrayToHeaderData; } });
__exportStar(require("./params.js"), exports);
__exportStar(require("./types.js"), exports);
